types/generate.js

```diff
@@ -0,0 +1,122 @@
+// TypeScript definitions are generated here.
+// AVA allows chaining of function names, like `test.after.cb.always`.
+// The order of these names is not important.
+// Writing these definitions by hand is hard. Because of chaining,
+// the number of combinations grows fast (2^n). To reduce this number,
+// illegal combinations are filtered out in `verify`.
+// The order of the options is not important. We could generate full
+// definitions for each possible order, but that would give a very big
+// output. Instead, we write an alias for different orders. For instance,
+// `after.cb` is fully written, and `cb.after` is emitted as an alias
+// using `typeof after.cb`.
+
+const path = require('path');
+const fs = require('fs');
+const runner = require('../lib/runner');
+
+const arrayHas = parts => part => parts.includes(part);
+
+const base = fs.readFileSync(path.join(__dirname, 'base.d.ts')).toString();
+
+// All suported function names
+const allParts = Object.keys(runner.chainableMethods).filter(name => name !== 'test');
+
+const output = base + generatePrefixed([]);
+fs.writeFileSync(path.join(__dirname, 'generated.d.ts'), output);
+
+// Generates type definitions, for the specified prefix
+// The prefix is an array of function names
+function generatePrefixed(prefix) {
+	let output = '';
+	let children = '';
+
+	for (const part of allParts) {
+		const parts = [...prefix, part];
+
+		if (prefix.includes(part) || !verify(parts, true)) {
+			// Function already in prefix or not allowed here
+			continue;
+		}
+
+		// Check that `part` is a valid function name.
+		// `always` is a valid prefix, for instance of `always.after`,
```

Comment:
Anything can be a function. Mix and match however you want.

`always.after` and `after.always` are the same thing.

You could technically do `always.always.always.after` - it would still be treated the same way.

The only thing disallowed is certain combinations:

Only one of `todo`, `skip`, `failing` should be allowed in the chain.

`always` only combines with the `after(Each)` hooks, etc.

Technically, the JavaScript allows any combination (you will never get a "not a function" TypeError) - we just validate the metadata to make sure the combinations make sense.